### ISSUE TITLE
Stale _gradIndex can collide across compiled graphs, throwing a transposed-shape error in AccumulateGrad

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#1022]
## Symptom

Training a model throws from the backward pass with a transposed shape pair:

```
System.ArgumentException : Tensor shapes must match. Got [256, 64] and [64, 256].
  at AiDotNet.Tensors.Engines.CpuEngine.TensorAddInPlace[T](Tensor`1 a, Tensor`1 b)
  at AiDotNet.Tensors.Engines.Autodiff.BackwardFunctions`1.MultiplyScalarBackward(...)
  at AiDotNet.Tensors.Engines.Autodiff.CompiledDelegateChain`1.Execute(...)
  at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradients(...)
```

Reproduced in AiDotNet on `CSDI<float>` (see ooples/AiDotNet#2091). `[256, 64]` and `[64, 256]` are the two feed-forward `DenseLayer` weights of the same block — `hiddenDim * 4 = 256` and `hiddenDim = 64` — so they have transposed shapes and **identical element counts**.

## Suspected mechanism: stale `_gradIndex` collision

Gradients are accumulated into a positional array rather than by tensor identity:

`DifferentiableOps.AccumulateGrad`:
```csharp
int idx = tensor._gradIndex;
if (idx >= 0 && _indexedGrads != null && idx < _indexedGrads.Length)
{
    var existing = (Tensor<T>?)_indexedGrads[idx];
    ...   // TensorAddInPlace(existing, grad)
}
```

Indices are assigned **only when unset** (`CompiledBackwardGraph.cs:163-169`):
```csharp
if (e.Output._gradIndex < 0) e.Output._gradIndex = gradIndexCount++;
if (e.Input0 != null && e.Input0._gradIndex < 0) e.Input0._gradIndex = gradIndexCount++;
```

and reset walks **only reachable entries** (`CompiledBackwardGraph.cs:~301`):
```csharp
foreach (int i in _reachableEntryIndices)
{
    e.Output._gradIndex = -1;
    ...
}
```

So a tensor that held an index in one compiled graph but is not in `_reachableEntryIndices` on a later pass keeps that stale index. On the next graph the `< 0` guard skips it, while `gradIndexCount` restarts from 0 and hands the same slot to a different tensor. Two tensors then accumulate into one slot — and when their shapes are transposes, `TensorAddInPlace` throws exactly as above.

Two weights with equal element counts and transposed shapes are the case most likely to surface it, because a slot collision between differently-sized tensors would fail differently or be masked.

**Confidence: this is a code-reading hypothesis, not yet proven by a minimal repro.** I have verified the assignment guard, the reachable-only reset, and that the two colliding shapes are the FFN weight pair. I have *not* yet demonstrated the stale index directly — that repro is the first thing to write.

## Aggravating factor

The reset does not appear to be guarded by `finally` (the enclosing `Execute` has `try`/`finally` blocks, but none within 40 lines above the reset). If the backward throws, the reset is skipped and every tensor keeps its index into the next call, so the fault would be self-perpetuating once triggered. This needs confirming by brace-depth analysis rather than proximity.

## Suggested next steps

1. A minimal repro: two parameters of transposed shape and equal element count, across two successive `ComputeGradients` calls where the second graph reaches a different entry set.
2. Assert in `AccumulateGrad` that `existing.Shape` matches `grad.Shape` before the in-place add, so a collision fails with a diagnostic naming both tensors rather than a bare shape error.
3. Consider whether the reset belongs in a `finally`, and whether index assignment should be scoped per-compilation rather than relying on a sentinel that survives across graphs.

Related: ooples/AiDotNet#2091

**Comment by ooples** [comment]
## Trigger isolated: `TensorArena` alone, by 2x2 factorial

Two factors had been observed together, and I previously reported the leading `Predict` as part
of the trigger. **That was wrong.** Running the four cells:

| arena | leading `Predict` | result |
|---|---|---|
| no | no | 8 steps OK |
| no | yes | 8 steps OK |
| **yes** | no | **FAILED at step 4** |
| **yes** | yes | **FAILED at step 3** |

`TensorArena` is necessary and sufficient. `Predict` is not necessary - it only moves the failure
one step earlier, which is what an accelerant looks like rather than a cause.

## Full frame

```
System.ArgumentException: Tensor shapes must match. Got [256, 64] and [64, 256].
  at AiDotNet.Tensors.Engines.CpuEngine.TensorAddInPlace[T](Tensor`1 a, Tensor`1 b)
  at AiDotNet.Tensors.Engines.Autodiff.BackwardFunctions`1.MultiplyScalarBackward(...)
  at AiDotNet.Tensors.Engines.Autodiff.CompiledDelegateChain`1.Execute(...)
  at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradientsViaGraphCore(...)
  at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradients(...)
```

`MultiplyScalarBackward` does one thing:

```csharp
var grad = engine.TensorMultiplyScalar(gradOutput, scalar);
DifferentiableOps.AccumulateGrad(grads, inputs[0], grad, engine);
```

so the tensor already holding a gradient for `inputs[0]` has the TRANSPOSED shape of the incoming
one. `[256, 64]` and `[64, 256]` are 16384 elements either way.

## Where the equal element count points

`TensorArena` pools by **size, not shape**:

```csharp
private readonly Dictionary<(Type, int), List<Array>> _pool = new();
```

A 16384-element `float[]` is therefore interchangeable between a `[256, 64]` tensor and a
`[64, 256]` one. That is exactly the signature observed: same element count, different shape,
only under an arena. A backing array released to the pool while still reachable as an accumulated
gradient would be reissued under the other shape on the next rent.

Stating the limit of the evidence: the factorial establishes the arena as the trigger and the pool
key explains why the two shapes are interchangeable. It does **not** yet identify which array is
returned early, and I have not bisected that. Two earlier hypotheses on this issue should be
treated as unproven:

- a stale `_gradIndex` surviving between passes - the assign and reset loops in
  `CompiledBackwardGraph` and `OptimizedBackwardPlan` both iterate the same reachable set, and
  `GradientTape` resets over all entries in a `finally`, so none of the three is asymmetric;
- a cached `CompiledDelegateChain` replayed against a graph it was not compiled for - the caller
  here opens a fresh `using var tape` per training step, so no chain persists across steps. The
  only state that does is the thread-static `BackwardScratch`.

## Repro

Consumer-side, in AiDotNet:

```csharp
using var arena = TensorArena.Create();          // remove this line and it passes
using var model = new CSDI<float>(new NeuralNetworkArchitecture<float>(
    inputType: InputType.OneDimensional,
    taskType: NeuralNetworkTaskType.Regression,
    inputSize: 512, outputSize: 4));

var input  = new Tensor<float>(new[] { 1, 512 });
var target = new Tensor<float>(new[] { 1, 4 });
for (int i = 0; i < input.Length;  i++) input[i]  = 0.01f * (i % 17);
for (int i = 0; i < target.Length; i++) target[i] = 0.1f * i;

for (int step = 1; step <= 8; step++)
    model.Train(input, target);                  // throws at step 4
```

CSDI is relevant only in that it samples a random timestep per call, so successive steps build
graphs with different reachable sets. Any model that varies its graph between steps under an
arena is a candidate for the same failure.
